src/hamster/session.py

```diff
@@ -0,0 +1,85 @@
+
+# Project Hamster is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with Project Hamster.  If not, see <http://www.gnu.org/licenses/>.
+
+import dbus
+import logging
+logger = logging.getLogger(__name__)
+
+class DbusSessionListener(object):
+    """Listen for GNOME manager end session event."""
+
+
+    def end_session(self):
+        """Override this method to do something at the end of a session.
+        It must not attempt to interact with the user and will be given
+        a maximum of ten seconds to perform the actions."""
+        pass
+
+
+    def __init__(self):
+        """Connect to the GNOME manager session signals"""
+
+        # Get SessionManager interface
+        session_bus = dbus.SessionBus()
+        try:
+            session_manager = session_bus.get_object("org.gnome.SessionManager",
+                "/org/gnome/SessionManager")
+            self.__session_manager_iface = dbus.Interface(session_manager,
+                dbus_interface="org.gnome.SessionManager")
+            self.__client_id = self.__session_manager_iface.RegisterClient("", "")
+
+            # Get SessionManager.ClientPrivate interface
+            session_client = session_bus.get_object("org.gnome.SessionManager",
+                self.__client_id)
+            self.__session_client_private_iface = dbus.Interface(session_client,
+                dbus_interface="org.gnome.SessionManager.ClientPrivate")
+
+            # Connect to the needed signals
+            session_bus.add_signal_receiver(self.__query_end_session_handler,
+                signal_name = "QueryEndSession",
+                dbus_interface = "org.gnome.SessionManager.ClientPrivate",
+                bus_name = "org.gnome.SessionManager")
+
+            session_bus.add_signal_receiver(self.__end_session_handler,
+                signal_name = "EndSession",
+                dbus_interface = "org.gnome.SessionManager.ClientPrivate",
+                bus_name = "org.gnome.SessionManager")
+
+            session_bus.add_signal_receiver(self.__stop_handler,
+                signal_name = "Stop",
+                dbus_interface = "org.gnome.SessionManager.ClientPrivate",
+                bus_name = "org.gnome.SessionManager")
+        except dbus.exceptions.DBusException:
+            logger.info("Unable to connect to GNOME session manager, stop tracking on logout won't work")
+
+    def __query_end_session_handler(self, flags):
+        """Inform that the session is about to end. It must reply with
+        EndSessionResponse within one second telling if it is ok to proceed
+        or not and why. If flags is true, the session will end anyway."""
+        self.__session_client_private_iface.EndSessionResponse(True, "")
```

Comment:
Why is this handler needed? If we're not doing anything other than saying it is ok to end the session, isn't this just the default? Or is it so that *if* you register something (I'm not sure about right terminology here) against SessionManager, that you *need* to also handle this QueryEndSession event?

Reply:
I have not checked, but according to the comment I have understood that when you register against the SessionManager you have to handle this message. 
If it's not done, I think it can delay the log out by 1 seconds, eventually the session manager could remove the corresponding program and we will not get the end session message.